File: Problem34.cpp
//34. Find First and Last Position of Element in Sorted Array - M.  Given an array of integers nums sorted in ascending order, find the starting and ending position of a given target value.
// Your algorithm's runtime complexity must be in the order of O(log n).
// If the target is not found in the array, return [-1, -1].

#include <iostream>
#include <vector>
using namespace std;

int extremeInsertionIndex(vector<int>& nums, int target, bool left) {
    int low = 0;
    int high = nums.size();

    while (low < high) {
        int mid = (low + high) / 2;
        if (nums[mid] > target || (left && target == nums[mid])) {
            high = mid;
        }
        else {
            low = mid+1;
        }
    }
    return low;
}

vector<int> searchRange(vector<int>& nums, int target) {
    vector<int> range = {-1, -1};

    int leftIdx = extremeInsertionIndex(nums, target, true);

    // assert that `leftIdx` is within the array bounds and that `target`
    // is actually in `nums`.
    if (leftIdx == nums.size() || nums[leftIdx] != target) {
        return range;
    }

    range[0] = leftIdx;
    range[1] = extremeInsertionIndex(nums, target, false)-1;

    return range;
}

int main() {
    vector<int> k;

    vector<int> temp1;
    k = searchRange(temp1, 0);
    for (int e : k){ cout<< e<< " ";}
    cout<<"\n";

    temp1.push_back(5);
    temp1.push_back(7);
    temp1.push_back(7);
    temp1.push_back(8);
    temp1.push_back(8);
    temp1.push_back(10);

    k = searchRange(temp1, 8);
    for (int e : k){ cout<< e<< " ";}
    cout<<"\n";

    vector<int> t2 {2,2};
    k = searchRange(t2, 1);
    for (int e : k){ cout<< e<< " ";}
    cout<<"\n";
    return 0;
}

File: NewJob2025/Problem34.java
// 34. Find First and Last Position of Element in Sorted Array - M
    public int searchRange(int[] nums, double target) {
        int low = 0;
        int high = nums.length-1;
        int mid = 0;

        while (low < high) {
            mid = low + (high-low)/2;
            if ((double) nums[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ((double) nums[low] > target) ? (low - 1) : (low);
    }

    public int[] searchRange(int[] nums, int target) {
        // we could technically find target-0.5, taget+0.5
        int[] empty = {-1, -1};
        if (nums.length == 0) return empty;
        if ((nums[0] > target) || (nums[nums.length-1] < target)) return empty;

        double lower = target - 0.5;
        double higher = target + 0.5;
        int lowerIdx = searchRange(nums, lower);
        int higherIdx = searchRange(nums, higher);
        if (lowerIdx == higherIdx) return empty;
        int[] res = {lowerIdx + 1, higherIdx};
        return res;
    }

// could use is_searching_left (a boolean indicating whether to search for the leftmost or rightmost occurrence of the target).
else if (nums[mid] == target)  {
                idx = mid;
                if (isSearchingLeft) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
